### DISCUSSION TITLE
help me curve

### DISCUSSION THREAD

**Discussion by 3donew** [ImageMagick/ImageMagick#7826] [Development]
### ImageMagick version

7.1.1

### Operating system, version and so on

Windows 11 x64

**Comment by 3donew** [comment]
help me create curve as shown in picture only in imagemagick. What command should i use?
![Снимок экрана 2024-12-08 003233](https://github.com/user-attachments/assets/7825c168-ced1-465e-9429-a61650a14432)

**Reply by fmw42** [reply]
Imagemagick is mostly command line driven.  So there is no GUI to do that. You can build a CLUT if you know the values at each point in your curve. That would be a 1D image that could then be used to do the same thing. see https://usage.imagemagick.org/color_mods/#color_lut. I have a bash unix script that could be used on Mac/Linux, but it does not work in Windows.

**Reply by 3donew** [reply]
> Imagemagick is mostly command line driven. So there is no GUI to do that. You can build a CLUT if you know the values at each point in your curve. That would be a 1D image that could then be used to do the same thing. see https://usage.imagemagick.org/color_mods/#color_lut. I have a bash unix script that could be used on Mac/Linux, but it does not work in Windows.

your script will write command for imagemagick? can you give me script? i will try

**Reply by fmw42** [reply]
It works only on Linux/Mac OSX/Windows Cygwin/Windows 10 Unix.  Not on normal Windows.

See http://www.fmwconcepts.com/imagemagick for script called curves

**Comment by snibgo** [comment]
I show many methods for making cluts at [Clut cookbook](http://im.snibgo.com/ckbkClut.htm). If you want certain pairs of input-output values, see the sections starting from "Cluts from splines from sparse values".

**Comment by 3donew** [comment]
here is my curve data, i need cli command for imagemagick that could change my image by this data
input 0; 10; 20; 40; 50; 60; 80; 100
output 43; 49; 56; 67; 76; 82; 87; 90

**Comment by 3donew** [comment]
![Снимок экрана 2024-12-08 230800](https://github.com/user-attachments/assets/756fc111-b65c-464d-8d36-d2562c0d96ba)

**Comment by snibgo** [comment]
Using one of the methods on the page I linked to:
```
magick ^
  -size 101x1 xc:None ^
  -fill graya(43%%,1) -draw "point 0,0" ^
  -fill graya(49%%,1) -draw "point 10,0" ^
  -fill graya(56%%,1) -draw "point 20,0" ^
  -fill graya(67%%,1) -draw "point 40,0" ^
  -fill graya(76%%,1) -draw "point 50,0" ^
  -fill graya(82%%,1) -draw "point 60,0" ^
  -fill graya(87%%,1) -draw "point 70,0" ^
  -fill graya(90%%,1) -draw "point 100,0" ^
  -define quantum:format=floating-point -depth 32 ^
  -write cl_spsecl.miff ^
  ( -clone 0 -alpha off -fill Black -colorize 100 ) ^
  ( -clone 0 -alpha extract -negate ) ^
  -compose seamless-blend ^
    -define compose:args=100000x1e-7+1000 ^
    -composite ^
  -alpha off ^
  clut.png
```

**Reply by 3donew** [reply]
> Using one of the methods on the page I linked to:
> 
> ```
> magick ^
>   -size 101x1 xc:None ^
>   -fill graya(43%%,1) -draw "point 0,0" ^
>   -fill graya(49%%,1) -draw "point 10,0" ^
>   -fill graya(56%%,1) -draw "point 20,0" ^
>   -fill graya(67%%,1) -draw "point 40,0" ^
>   -fill graya(76%%,1) -draw "point 50,0" ^
>   -fill graya(82%%,1) -draw "point 60,0" ^
>   -fill graya(87%%,1) -draw "point 70,0" ^
>   -fill graya(90%%,1) -draw "point 100,0" ^
>   -define quantum:format=floating-point -depth 32 ^
>   -write cl_spsecl.miff ^
>   ( -clone 0 -alpha off -fill Black -colorize 100 ) ^
>   ( -clone 0 -alpha extract -negate ) ^
>   -compose seamless-blend ^
>     -define compose:args=100000x1e-7+1000 ^
>     -composite ^
>   -alpha off ^
>   clut.png
> ```

I don't understand how to use this

**Reply by 3donew** [reply]
found on the forum how to use fx
here is an example magick lena_gray.png -fx "uu=1.127*u*u - 0.3851*u + 0.0614" lena_gray_fx3.png
this method would suit me only how to get the data that I need

**Comment by snibgo** [comment]
> I don't understand how to use this

My command makes a clut file. Use it like this:
```
magick in.png clut.png -clut out.png
```
This reads in.png, transforms it according to the curve, writing to out.png.

**Reply by 3donew** [reply]
> > I don't understand how to use this
> 
> My command makes a clut file. Use it like this:
> 
> ```
> magick in.png clut.png -clut out.png
> ```
> 
> This reads in.png, transforms it according to the curve, writing to out.png.

Thanks. It worked for me.

**Comment by snibgo** [comment]
> found on the forum how to use fx

You can find the three parameters by quadratic regression, eg https://www.omnicalculator.com/statistics/quadratic-regression
